alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+
+use winapi::shared::ntdef::{BOOLEAN, HANDLE, PVOID};
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback for `HandleWaitSignal`, unpacks Rust closure reference and calls it
+extern "system" fn child_exit_callback<F>(ctx: PVOID, timed_out: BOOLEAN)
+where
+    F: FnOnce() + Send,
+{
+    if timed_out != 0 {
+        return;
+    }
+
+    let callback: Box<F> = unsafe { Box::from_raw(ctx as *mut F) };
+    callback();
+}
+
+/// Represents closure attached to Win32 handle wait signal
+///
+/// This allows to fire a Rust callback when subprocess exits
+pub(crate) struct HandleWaitSignal {
+    wait_handle: AtomicPtr<c_void>,
+}
+
+impl HandleWaitSignal {
+    /// Registers an asynchronous closure to call when `child_handle` wait signals.
+    ///
+    /// The `on_exit` is called on Win32 threadpool thread so it should avoid
+    /// blocking calls. See [`WT_EXECUTEINWAITTHREAD` flag docs] for details
+    ///
+    /// [`WT_EXECUTEINWAITTHREAD` flag docs]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject
+    fn new<F>(child_handle: HANDLE, on_exit: F) -> Result<HandleWaitSignal, Error>
+    where
+        F: FnOnce() + Send,
+    {
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let wait_notify = Box::new(on_exit);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub(crate) struct ChildProcessState {
+    _on_exit: HandleWaitSignal,
+    events: Receiver<ChildEvent>,
+}
+
+impl ChildProcessState {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessState, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            if let Err(e) = sender.send(ChildEvent::Exited) {
+                warn!(
+                    "An error occurred while attempting to notify about child process z
+                    termination: {}",
+                    e
+                );
+            }
+        })?;
+
+        Ok(ChildProcessState { _on_exit: on_exit, events: receiver })
+    }
+
+    pub fn events(&self) -> &Receiver<ChildEvent> {
```

Comment:
Why is this named `events`? Makes it sound like it gives back a `Vec<Event>` or something like that.

If there's no deeper meaning behind that, we might want to rename it to something that says more clearly that this is a receiving channel. I personally found `child_state.events()` a bit hard to understand.

Reply:
It's prepared to be used with [`EventedPty`](https://github.com/jwilm/alacritty/blob/master/alacritty_terminal/src/tty/mod.rs#L68) in case [`ChildEvent`](https://github.com/jwilm/alacritty/blob/master/alacritty_terminal/src/tty/mod.rs#L58) enum gets more cases.

Reply:
It still doesn't return a list of events, but a receiver. That's a significant difference that should usually be exposed at the callee site. Something like `event_rx` would probably make this more readable.